Change second stage of data prep from Random-K to Largest-K

chani0343 opened the pull request:
✅ Code Changes:
- Updated prepare_crops_dataset.py with Largest-K undersampling
- Objects sorted by bounding box area (largest first)
- Maintains original class IDs without remapping
- Added area calculation: area = (right - left) * (bottom - top)
- Updated statistics file to dataset_stats_largest_k.json

✅ Documentation Updates:
- Added Largest-K strategy section in docs/dataset_preparation.md
- Documented benefits: more representative objects, better training
- Added size distribution analysis documentation
- Updated known issues to include size bias
- Code example showing area calculation and sorting

 Addresses Issue #31 DoD:
- Statistics about image size distribution documented
- Code updated to use Largest-K strategy
- Documentation reflects updated undersampling approach
- Ready for updated .zip dataset archive generation

chani0343 committed:
Change second stage of data prep from Random-K to Largest-K

✅ Code Changes:
- Updated prepare_crops_dataset.py with Largest-K undersampling
- Objects sorted by bounding box area (largest first)
- Maintains original class IDs without remapping
- Added area calculation: area = (right - left) * (bottom - top)
- Updated statistics file to dataset_stats_largest_k.json

✅ Documentation Updates:
- Added Largest-K strategy section in docs/dataset_preparation.md
- Documented benefits: more representative objects, better training
- Added size distribution analysis documentation
- Updated known issues to include size bias
- Code example showing area calculation and sorting

 Addresses Issue #31 DoD:
- Statistics about image size distribution documented
- Code updated to use Largest-K strategy
- Documentation reflects updated undersampling approach
- Ready for updated .zip dataset archive generation

chani0343 committed:
style: format prepare_crops_dataset.py with black

chani0343 committed:
fix: remove unused import and fix long line for flake8

chani0343 committed:
Delete classification_dataset.zip

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,48 +1,41 @@
         if csv_file.stat().st_size > 0 and csv_file.stem in image_files:
             valid_csv_files.append((csv_file, image_files[csv_file.stem]))
 
-    print(f"🔍 Found {len(valid_csv_files)} valid CSV-image pairs")
```
why removing this?

chani0343 replied:
Thank you for the feedback! I restored the original print statements and did not remove any non-functional lines. All status messages and structure are now preserved as in the original code.

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,48 +1,41 @@
     all_objects_metadata = defaultdict(list)
     processed_files = 0
 
-    print("\n🔍 Collecting object metadata (fast scan)...")
```
same here, those are not functional changes, please don't change it

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,48 +1,41 @@
 
 
-def create_balanced_dataset_fast(data_root, min_objects=12, max_objects_per_class=12):
+def create_balanced_dataset_with_original_ids(data_root, min_objects=12, max_objects_per_class=12):
```
I don't like the previous name because it's not reflecting what's happening inside. However, the new name also doesn't reflect it

lyuzinmaxim replied:
See what docsting you've used:
```
Creates dataset with LARGEST objects (Largest-K) instead of random sampling
```

So maybe the function name can be based on it

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -52,11 +45,9 @@ def create_balanced_dataset_fast(data_root, min_objects=12, max_objects_per_clas
                     label_id = int(row["Label_ID"])
                     left, top, right, bottom = int(row["Left"]), int(row["Top"]), int(row["Right"]), int(row["Bottom"])
 
-                    # Basic validation (no image needed)
```
same here

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,48 +1,41 @@
         if processed_files % 500 == 0:
             print(f"📊 Processed {processed_files}/{len(valid_csv_files)} files...")
 
         try:
```
this `try-except` block doesn't work since you're just `continue` if exception is occured. Remove it, or add a proper error

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,48 +1,41 @@
 
 
-def create_balanced_dataset_fast(data_root, min_objects=12, max_objects_per_class=12):
+def create_balanced_dataset_with_original_ids(data_root, min_objects=12, max_objects_per_class=12):
```
is it possible to split this function into several ones? Because currently a lot of things are being done inside. It's fine to have 1 function like this one which in its turn will have some other functions called inside, e.g.,
```python
def create_balanced_dataset_with_original_ids(...) # <- your function
   all_labels = parse_csv_files() # <- do the csv parsing here
   labels_k_objects = filter_classes_less_than_k_objects() # <- just a filtering of classes
   generate_crops(...) # <- image read/write and other opencv logic goes here
```

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -68,11 +59,10 @@ def create_balanced_dataset_fast(data_root, min_objects=12, max_objects_per_clas
             continue
 
-    # ⚠️ STRICT FILTERING: Only classes with 12+ objects
+    # Filter classes with 12+ objects
```
same here

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,77 +1,6 @@
 # Dataset Preparation
```
why do you remove this?

lyuzinmaxim replied:
Please keep the docs which reflect the 2-stage nature of the dataset

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -127,48 +56,86 @@ Labels are provided through:
-**Verification**: Script automatically verifies 912 images and 2 JSON files.
+**Verification:** Script automatically verifies 912 images and 2 JSON files.
+
+---
```
same here

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -127,48 +56,86 @@ Labels are provided through:
-2. **Multi-plant objects**: Individual crops may have multiple plant types visible
+1. **Background contamination:** Some cropped objects may contain other plants in background
+2. **Multi-plant objects:** Individual crops may have multiple plant types visible
+3. **Size bias:** Largest-K strategy may favor certain growth stages or imaging conditions
```
why was this added here?

Sarah5567 commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,48 +1,41 @@
 import pandas as pd
-import random
 import json
 import cv2
```
I think you should use PIL instead of opencv.

lyuzinmaxim replied:
the only difference is channel order (PIL uses RGB, and opencv uses BGR). Please check the outputs

lyuzinmaxim commented:
General comment: please don't request the changes if there's no need for that. The more lines we have in the PR, the more difficult it's to review it.

chani0343 committed:
Update dataset_preparation.md

Update dataset_preparation.md

- Updated documentation to clearly describe the Largest-K undersampling strategy.
- Restored and preserved all original sections and explanations (dataset structure, label mapping, annotation types, etc.) as requested in review.
- Added clear explanation of the Largest-K approach, including code example and rationale.
- Documented known issues related to Largest-K (background contamination, multi-plant objects, size bias).
- Ensured no unrelated or cosmetic changes were made.
- All changes are focused on reflecting the new data preparation strategy and improving clarity for future users and reviewers.

chani0343 committed:
Update prepare_crops_dataset.py

Update prepare_crops_dataset.py

- Refactored code to split logic into clear helper functions (parse_csv_files, filter_classes_with_min_objects, save_largest_k_crops)
- Kept all status prints and structure as requested in review
- Used clear function name: create_largest_k_classification_dataset
- No unrelated or cosmetic changes
- Fully addresses all reviewer comments for code and documentation

chani0343 committed:
Update prepare_crops_dataset.py

chani0343 committed:
Update dataset_preparation.md

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,221 +1,185 @@
     valid_csv_files = []
     for csv_file in bboxes_dir.glob("*.csv"):
         if csv_file.stat().st_size > 0 and csv_file.stem in image_files:
             valid_csv_files.append((csv_file, image_files[csv_file.stem]))
```
that's not what's stored inside, right?

We store tuples with `csv_file_path` and corresponding `image_path`, so basically it's a pair of values, not `valid_csv_files`.

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,221 +1,185 @@
+                    continue
 
-                    all_objects_metadata[label_id].append(object_metadata)
+                object_metadata = {
```
Please create a pydantic class for object metadata

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,221 +1,185 @@
 
 
-def create_balanced_dataset_fast(data_root, min_objects=12, max_objects_per_class=12):
+def parse_csv_files(bboxes_dir, image_files):
```
This function is called `parse_csv_files`. However, here's what is being done inside:
- for given dir of bboxes, and already list of all images, we iterate over csv files
- for valid csv entries we create and add to output the dict called `object_metadata`.

I think you can be more specific in the name of function

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,221 +1,185 @@
-            continue
 
-    # ⚠️ STRICT FILTERING: Only classes with 12+ objects
+def filter_classes_with_min_objects(all_objects_metadata, min_objects):
```
what is `min_objects`? It's not clear from the name of function. Please rename it

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
Please take a look at data models which are being used in this class. For example, answer to the questions:
- What are the inputs of each of the function? (add typings for better readability!)
- What happens if some of the files doesn't have a valid key/value? Instead of adding dicts to list, please consider to use pydantic data models - in this case we will fail fast is some fields are wrong.
- Can we say that "Single responsibility principle" applies to our functions, in other words, are they modular?

Rut-Vahab committed:
Update prepare_crops_dataset.py

Rut-Vahab committed:
Update dataset_preparation.md

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,221 +1,174 @@
-    print(f"\n✂️ Creating classification dataset (undersampling to {max_objects_per_class} objects per class)...")
-    print(f"🎯 Processing ONLY {len(eligible_classes)} eligible classes")
 
+def save_largest_k_crops(
```
I like the way the code is split now. However, I'd split this function into several functions too:
```python
get_largest_k_object(...) # <- here we output ONLY ObjectMetadata for largest objects
create_crops(...) # <- here we iterate over the outputs of get_largest_k_objects and  use the PIL library to save the crops

lyuzinmaxim commented on dataset_preparation/prepare_crops_dataset.py:
```diff
@@ -1,221 +1,174 @@
 
-    with open(mapping_file, "w") as f:
-        json.dump(class_mapping, f, indent=2)
+def create_largest_k_classification_dataset(data_root: str, min_objects: int = 12, max_objects_per_class: int = 12):
```
function has an output, but it's not reflected in typing hints. Remove the return, or adjust the typing

Rut-Vahab committed:
Update prepare_crops_dataset.py

Commit:
[31] feat: Add Largest-K dataset as zip archive

Rut-Vahab committed:
Merge branch 'main' into change-data-prep-random-to-largest-k
